libnopegl/src/text.c

```diff
@@ -127,6 +138,220 @@ int ngli_text_init(struct text *s, const struct text_config *cfg)
     return s->cls->init(s);
 }
 
+void ngli_text_update_effects_defaults(struct text *s, const struct text_effects_defaults *defaults)
+{
+    s->config.defaults = *defaults;
+}
+
+static int set_value_from_node(float *dst, struct ngl_node *node, double t)
+{
+    int ret = ngli_node_update(node, t);
+    if (ret < 0)
+        return ret;
+    const struct variable_info *v = node->priv_data;
+    memcpy(dst, v->data, v->data_size);
+    return 0;
+}
+
+static int set_f32_value(float *dst, struct ngl_node *node, float value, double t)
+{
+    if (node)
+        return set_value_from_node(dst, node, t);
+    if (value < 0.f)
+        return 0;
+    *dst = value;
+    return 0;
+}
+
+static int set_vec3_value(float *dst, struct ngl_node *node, const float *value, double t)
+{
+    if (node)
+        return set_value_from_node(dst, node, t);
+    if (value[0] < 0.f)
+        return 0;
+    memcpy(dst, value, 3 * sizeof(*dst));
+    return 0;
+}
+
+static int set_transform(float *dst, struct ngl_node *node, double t)
+{
+    if (!node)
+        return 0;
+    int ret = ngli_node_update(node, t);
+    if (ret < 0)
+        return ret;
+    ngli_transform_chain_compute(node, dst);
+    return 0;
+}
+
+static void reset_chars_data_to_defaults(struct text *s)
+{
+    memcpy(s->chars_data, s->chars_data_default, s->chars_data_size);
+}
+
+static struct text_effects_pointers get_chr_data_pointers(float *base, size_t nb_chars)
+{
+    struct text_effects_pointers ptrs = {0};
+    ptrs.transform  = base;
+    ptrs.color      = ptrs.transform  + nb_chars * 4 * 4;
+    ptrs.opacity    = ptrs.color      + nb_chars * 3;
+    return ptrs;
+}
+
+static void segment_chars(const struct text *s, struct effect_segmentation *effect)
+{
+    size_t char_id = 0;
+    size_t position = 0;
+
+    const struct char_info_internal *chars = ngli_darray_data(&s->chars_internal);
+    for (size_t i = 0; i < ngli_darray_count(&s->chars_internal); i++) {
+        const struct char_info_internal *c = &chars[i];
+        if (c->tags & NGLI_TEXT_CHAR_TAG_GLYPH)
+            effect->positions[char_id++] = position;
+        position++; // account for all the hidden characters as if they existed
```

Comment:
as if they exist

Reply:
You sure about that? 🤔

Reply:
Mmm, according to google trad (for what it is worth), it translate `existed`/`exist` to the same tense in french and when going back from french to english, it uses `exist`.